fix issues with the java bytecode frontend

Timbals opened the pull request:
Sorry for the big PR. The bytecode fronted has a lot of shared state that makes it difficult to change in small steps.

Moved the logic for assigning the values of operands to stack locals out of `OperandStack` and into `Operand`.
That means that the assignments are created when the operands are used as a local/immediate with the `toLocal` and `toImmediate` methods instead of when they are popped of the stack with `popLocal` and `popImmediate`.
This fixes some issues where the wrong `pop` method was used which caused unnecessary stack local variables to get created.
This also results in improved type-safety and gets rid of a bunch of casts, because `toLocal` and `toImmediate` actually return the `Local` and `Immediate` types instead of a generic `Operand`.

Significantly simplified merging logic in `StackFrame`.

All `convert` methods had a code path for the first time they were called and for merging subsequent calls.
These two code paths have been combined now.

Factored a bunch of duplicated code into the new `changeStackLocal` in `Operand`.
The old version actually incorrectly merged operands at times (See `SwitchExprWithYieldTest`).

This code
```java
int k = 1;
String s = "";

switch (k) {
  case 1:
    s += "single";
  default: // note the fall-through
    s += "somethingElse";
};
```
would previously produce
```jimple
$l1 = 1
$l2 = ""
$stack8 = $l1
switch($stack8)
case 1: goto label13
default: goto label14
label13:
$l2 = dynamicinvoke "makeConcatWithConstants" <java.lang.String (java.lang.String)>($l2) <java.lang.invoke.StringConcatFactory: java.lang.invoke.CallSite makeConcatWithConstants(java.lang.invoke.MethodHandles$Lookup,java.lang.String,java.lang.invoke.MethodType,java.lang.String,java.lang.Object[])>("\u0001single")
label14:
$stack4 = $l2
$stack5 = dynamicinvoke "makeConcatWithConstants" <java.lang.String (java.lang.String)>($l2) <java.lang.invoke.StringConcatFactory: java.lang.invoke.CallSite makeConcatWithConstants(java.lang.invoke.MethodHandles$Lookup,java.lang.String,java.lang.invoke.MethodType,java.lang.String,java.lang.Object[])>("\u0001somethingElse")
// there is a bug here: `$l2 = $stack5` is missing
```
but now produces
```jimple
$l1 = 1
$l2 = ""
switch($l1)
case 1: goto label13
default: goto label14
label13:
$l2 = dynamicinvoke \"makeConcatWithConstants\" <java.lang.String (java.lang.String)>($l2) <java.lang.invoke.StringConcatFactory: java.lang.invoke.CallSite makeConcatWithConstants(java.lang.invoke.MethodHandles$Lookup,java.lang.String,java.lang.invoke.MethodType,java.lang.String,java.lang.Object[])>(\"\\u0001single\")
label14:
$l2 = dynamicinvoke \"makeConcatWithConstants\" <java.lang.String (java.lang.String)>($l2) <java.lang.invoke.StringConcatFactory: java.lang.invoke.CallSite makeConcatWithConstants(java.lang.invoke.MethodHandles$Lookup,java.lang.String,java.lang.invoke.MethodType,java.lang.String,java.lang.Object[])>(\"\\u0001somethingElse\")
```
This adds the missing assignment and doesn't produce so many stack local variables.

Fixed updating usages of the `value`/`stackLocal` in `Operand`. This is now part of the `changeStackLocal` method.

Simplified duplication instructions (Thanks to `Operand.DWORD_DUMMY`).

Removed some commented out code.

Fixed an issue in the `ReplaceUseStmtVisitor` that would produce stale results.

Fixes #698
Fixes #631 

TODO:
- [X] rebase/merge to newest develop
- [X] fix tests
- [X] add comments/documentation on how the bytecode frontend works for the next person that looks at this code
- [X] don't create stack locals when a normal local could be used
- [X] rename/rework `StackFrame` (it has nothing to do with stack frames)
- [X] remove duplicated logic in *each* instruction conversion method for creating and merging the instruction
- [X] don't merge operands when they all have the same value
- [X] find out what `addReadOperandAssignments_internal` does ~~and maybe remove it~~
- [X] improve statement position info for operands and statements

Timbals committed:
fix stale result in `ReplaceUseStmtVisitor`

Certain assign statement would produce in no result. Because the visitor is stateful, this would result in an old result being used later.

Timbals committed:
fix issues with the java bytecode frontend

Moved the logic for assigning the values of operands to stack locals out of `OperandStack` and into `Operand`.
That means that the assignments are created when the operands are used as a local/immediate with the `toLocal` and `toImmediate` methods instead of when they are popped of the stack with `popLocal` and `popImmediate`.
This fixes some issues where the wrong `pop` method was used which caused unnecessary stack local variables to get created.
This also results in improved type-safety and gets rid of a bunch of casts, because `toLocal` and `toImmediate` actually return the `Local` and `Immediate` types instead of a generic `Operand`.

Significantly simplified merging logic in `StackFrame`.
Factored a bunch of duplicated code into the new `changeStackLocal` in `Operand`.
The old version actually incorrectly merged operands at times (See `SwitchExprWithYieldTest`). I will add a test for this in a later commit.

Fixed updating usages of the `value`/`stackLocal` in `Operand`. This is now part of the `changeStackLocal` method.

Simplified duplication instructions (Thanks to `Operand.DWORD_DUMMY`).

Removed some commented out code.

Timbals committed:
don't merge operands when their values match

Timbals committed:
start inlining STORE instructions

Timbals committed:
add test for merging with duplicate values

Timbals committed:
fix tests for bytecode frontend changes

virusdefender commented:
Will you rebase the develop branch to the current branch recently? I want to try to validate the changes, but I found a lot of conflicts

Timbals committed:
unify "first time" and "merge" code paths in all convert methods

This removes the duplicated logic that was present in *every* convert method with any inputs/outputs.

Timbals committed:
document `addReadOperandAssignments`

Timbals committed:
rename `StackFrame` to `OperandMerging`

The class had nothing to do with stack frames

Timbals committed:
add more comments

Timbals committed:
remove remaining duplicated code paths

Timbals committed:
don't replace the operand value when creating a new stack local when merging

Instead, rely on the fact that all `convert` methods override their respective statements.
That means just changing the incoming operands will automatically correctly update the statement.

Also added a test case to make sure this works with the special case of using a normal local as a stack local.

Timbals committed:
fix position info for operands

Timbals committed:
fix position info for statements

When converting the instructions for a branch, there might not be a line number node after the branch target.
This would result in incorrect line numbers being used until the next line number node.

The fix is to simply store the line number for branches and restore the current line number when handling that branch.

swissiety reviewed (approved):
looks excellent - thx for cleaning that up :heart:
